x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/common/notifications/AbstractAuditor.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.core.common.notifications;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.index.IndexRequest;
+import org.elasticsearch.action.index.IndexResponse;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Objects;
+
+import static org.elasticsearch.common.xcontent.XContentFactory.jsonBuilder;
+import static org.elasticsearch.xpack.core.ClientHelper.executeAsyncWithOrigin;
+
+public abstract class AbstractAuditor<T extends AbstractAuditMessage> {
```

Comment:
I wonder if this needs to be `abstact`. We could pass `executionOrigin` and `auditIndex` in the constructor. And we could simply assume all implementation would simply log-trace success/failure, or at least those 2 could have a default implementation which most implementors would not need override.

Reply:
@dimitris-athanasiou I waffled back and forth while writing this on it being abstract or not. Comes down to inheritance vs injection. I am fine with simply passing the appropriate strings if that is preferred. Especially since nothing really interacts with the `AbstractAuditor` type by-itself.